File: tuilab/sdks/javascript/src/binary.ts
/** Locate the `tuilab` binary: explicit → TUILAB_BIN → @tui-lab/cli → workspace → PATH. */import { existsSync } from "node:fs";
import * as path from "node:path";
import { TuiLabError } from "./errors.js";

function platformKey(): string {
  const arch = process.arch === "arm64" ? "arm64" : "x64";
  return `${process.platform}-${arch}`;
}

/** Engine bundled via the @tui-lab/cli platform package, if installed. */
function cliPackageBinary(): string | null {
  const file = process.platform === "win32" ? "tuilab.exe" : "tuilab";
  try {
    const pkg = require.resolve(`@tui-lab/cli-${platformKey()}/package.json`, {
      paths: [__dirname],
    });
    const candidate = path.join(path.dirname(pkg), file);
    return existsSync(candidate) ? candidate : null;
  } catch {
    return null;
  }
}

export function findBinary(explicit?: string): string {
  if (explicit) {
    if (existsSync(explicit)) return explicit;
    throw new TuiLabError(`tuilab binary not found: ${explicit}`);
  }
  const env = process.env["TUILAB_BIN"];
  if (env && existsSync(env)) return env;
  // Installed platform package wins over PATH: a PATH `tuilab` may be the
  // @tui-lab/cli node launcher, which must never be spawned as the engine.
  // (The launcher itself resolves the platform package directly, so no loop.)
  const packaged = cliPackageBinary();
  if (packaged) return packaged;
  // dist/binary.js → sdks/javascript → sdks → tuilab → target/debug
  const workspace = path.resolve(
    __dirname,
    "..",
    "..",
    "..",
    "target",
    "debug",
    process.platform === "win32" ? "tuilab.exe" : "tuilab"
  );
  if (existsSync(workspace)) return workspace;
  for (const dir of (process.env["PATH"] ?? "").split(path.delimiter)) {
    const candidate = path.join(
      dir,
      process.platform === "win32" ? "tuilab.exe" : "tuilab"
    );
    if (existsSync(candidate)) return candidate;
  }
  throw new TuiLabError(
    "tuilab binary not found: set TUILAB_BIN, add it to PATH, " +
      "or build the workspace (cargo build -p tui-lab-cli)"
  );
}

File: tuilab/sdks/rust/src/binary.rs
//! Locate the `tuilab` binary: explicit → TUILAB_BIN → PATH → workspace.

use std::path::{Path, PathBuf};

use crate::error::TuiLabError;

/// Resolve the `tuilab` binary.
pub fn find_binary(explicit: Option<&Path>) -> Result<PathBuf, TuiLabError> {
    if let Some(path) = explicit {
        if path.is_file() {
            return Ok(path.to_path_buf());
        }
        return Err(TuiLabError::new(format!(
            "tuilab binary not found: {}",
            path.display()
        )));
    }
    if let Ok(from_env) = std::env::var("TUILAB_BIN") {
        let path = PathBuf::from(&from_env);
        if path.is_file() {
            return Ok(path);
        }
    }
    if let Some(path) = search_path() {
        return Ok(path);
    }
    // src files live at sdks/rust/src → walk up to tuilab/ → target/debug.
    if let Ok(manifest) = std::env::var("CARGO_MANIFEST_DIR") {
        let candidate = PathBuf::from(manifest)
            .join("..")
            .join("..")
            .join("target")
            .join("debug")
            .join(format!("tuilab{}", std::env::consts::EXE_SUFFIX));
        if candidate.is_file() {
            return Ok(candidate);
        }
    }
    Err(TuiLabError::new(
        "tuilab binary not found: set TUILAB_BIN, add it to PATH, \
         or build the workspace (cargo build -p tui-lab-cli)",
    ))
}

/// Search `PATH` for the binary (plus `.exe` on Windows).
fn search_path() -> Option<PathBuf> {
    let name = format!("tuilab{}", std::env::consts::EXE_SUFFIX);
    std::env::var_os("PATH").and_then(|paths| {
        std::env::split_paths(&paths)
            .map(|dir| dir.join(&name))
            .find(|candidate| candidate.is_file())
    })
}
